### PR TITLE
Configure byte-buddy to use simpler method-graph

### PR DESCRIPTION
**Pull request by mcculls**
# What Does This Do

Configures byte-buddy to use a simpler method graph compiler which just considers locally declared methods. This approach is enough for our transformations which insert advice into methods. And since our transformations use either `@Advice` or `AsmVisitorWrapper` we can also turn off generation of visibility bridges and use a simpler type strategy.

To disable this optimization add this JVM option:
```
-Ddd.resolver.simple.method.graph=false
```
or set this environment variable:
```
DD_RESOLVER_SIMPLE_METHOD_GRAPH=false
```

# Motivation

For spring-petclinic this reduces the number of class-file lookups by a fifth.

OTel uses a similar configuration for byte-buddy: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/v2.0.0/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java#L140

# Additional Notes

We don't enable the simpler method graph compiler when creating GraalVM native images because it appears to affect the reachability analysis, resulting in intermittent warnings like `Discovered unresolved field during parsing`.

Jira ticket: [APMS-11443]

[APMS-11443]: https://datadoghq.atlassian.net/browse/APMS-11443?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/testing/src/test/java/locator/ClassInjectingTransformer.java [hunk 1]
```diff
@@ -59,6 +59,7 @@ public static void injectInterfaceNamed(String binaryName, ClassLoader classLoad
           null,
           "java/lang/Object",
           null);
+      cw.visitAnnotation("Ljava/lang/FunctionalInterface;", true).visitEnd();
```
Comment (PerfectSlayer) [inline dd-java-agent/testing/src/test/java/locator/ClassInjectingTransformer.java:62]: Just curious: what's the reason behind adding the `@FunctionalInterface`?

**Review by PerfectSlayer — APPROVED** [review]
Looks good!
Thanks for adding a lot of details and context, it definitely helps to review 🙏

File: dd-java-agent/testing/src/test/java/locator/ClassInjectingTransformer.java (hunk 1 above)
Comment (mcculls) [reply dd-java-agent/testing/src/test/java/locator/ClassInjectingTransformer.java:62]: Previously this test relied on the instrumentation failing to apply just because there was a missing type somewhere in the class hierarchy (even if that type might never be used by any matchers or the advice) - with the simpler method graph compiler the old matcher no longer triggers an exception, so this test was failing - even though what it was trying to test, ie. that the `defineClass` fallback was not being used, was still true.

This PR changes the test to instead make the inserted type have an annotation (it can be any annotation) and then match against that annotation - in order to check if the type is annotated, we'll need to load it, which we cannot do when the `defineClass` fallback is turned off (since this class is deliberately generated dynamically so there's no class file backing it.)

TLDR; we're adjusting the test matcher to explicitly depend on loading the type, so absence of it will fail.

The use of `@FunctionalInterface` is just because it's available - any annotation would do.
